### PR TITLE
feat: separate DAL for Prover and Main server

### PR DESCRIPTION
**Pull request by AnastasiiaVashchuk**
Todo: 
+ update docs

### CODE REVIEW

File: core/lib/db_utils/src/instrument.rs [hunk 1]
```diff
@@ -238,43 +238,3 @@ where
...
-
-        let mut conn = pool.access_storage().await.unwrap();
-        sqlx::query("SELECT pg_sleep(1.5)")
-            .map(drop)
-            .instrument("slow")
-            .with_arg("miniblock", &MiniblockNumber(1))
-            .with_arg("hash", &H256::zero())
-            .fetch_optional(conn.conn())
-            .await
-            .unwrap();
```
Comment (AnastasiiaVashchuk) [inline core/lib/db_utils/src/instrument.rs:278]: these tests are moved to tests in `server_dal` crate

File: core/lib/server_dal/src/blocks_dal.rs [hunk 2]
```diff
@@ -1308,64 +1308,6 @@ impl BlocksDal<'_, '_> {
...
-        )
-        .fetch_all(self.storage.conn())
-        .await?;
-
-        Ok(rows
-            .into_iter()
-            .map(|row| (row.l1_batch_number, row.merkel_tree_paths_blob_url.unwrap()))
-            .collect())
-    }
-
```
Comment (AnastasiiaVashchuk) [inline core/lib/server_dal/src/blocks_dal.rs:1368]: https://github.com/matter-labs/zksync-era/pull/369/files#r1388250810

File: core/lib/zksync_core/src/proof_data_handler/mod.rs [hunk 3]
```diff
@@ -34,7 +34,7 @@ pub(crate) async fn run_server(
     config: ProofDataHandlerConfig,
     contracts_config: ContractsConfig,
     blob_store: Box<dyn ObjectStore>,
-    pool: ConnectionPool,
+    pool: ServerConnectionPool,
```
Comment (Deniallugo) [inline core/lib/zksync_core/src/proof_data_handler/mod.rs:37]: proof data handler with server?

File: core/lib/zksync_core/src/witness_generator/basic_circuits.rs [hunk 4]
```diff
@@ -252,16 +256,24 @@ impl JobProcessor for BasicWitnessGenerator {
...
     started_at: Instant,
     block_number: L1BatchNumber,
     job: PrepareBasicCircuitsJob,
 ) -> BasicCircuitArtifacts {
-    let witness_gen_input =
-        build_basic_circuits_witness_generator_input(connection_pool.clone(), job, block_number)
-            .await;
-    let (basic_circuits, basic_circuits_inputs, scheduler_witness) =
-        generate_witness(object_store, config, connection_pool, witness_gen_input).await;
+    let witness_gen_input = build_basic_circuits_witness_generator_input(
```
Comment (Deniallugo) [inline core/lib/zksync_core/src/witness_generator/basic_circuits.rs:264]: @EmilLuta could you double that it's the correct pool

File: docs/architecture.md [hunk 5]
```diff
@@ -51,7 +51,7 @@ This section provides a physical map of folders & files in this repository.
     - `/config`: All the configured values used by the different zkSync apps.
     - `/contracts`: Contains definitions of commonly used smart contracts.
     - `/crypto`: Cryptographical primitives used by the different zkSync crates.
-    - `/dal`: Data availability layer
+    - `/server_dal`: Data availability layer
```
Comment (Deniallugo) [inline docs/architecture.md:54]: we have to include prover dal somewhere as well

File: etc/env/base/rust.toml [hunk 6]
```diff
@@ -9,7 +9,7 @@ zksync_core=debug,\
 zksync_server=debug,\
 zksync_prover=debug,\
 zksync_contract_verifier=debug,\
-zksync_dal=info,\
+zksync_server_dal=info,\
```
Comment (Deniallugo) [inline etc/env/base/rust.toml:12]: add prover_dal as well

File: etc/system-contracts
Comment (Deniallugo) [inline etc/system-contracts:1]: please don't change it

File: prover/witness_generator/src/main.rs [hunk 7]
```diff
@@ -150,7 +154,7 @@ async fn main() -> anyhow::Result<()> {
                 config,
                 &store_factory,
                 public_blob_store,
-                connection_pool,
+                server_connection_pool,
```
Comment (Deniallugo) [inline prover/witness_generator/src/main.rs:157]: prover?

File: prover/witness_generator/src/basic_circuits.rs [hunk 8]
```diff
@@ -107,8 +107,8 @@ impl BasicWitnessGenerator {
 
     async fn process_job_impl(
         object_store: Arc<dyn ObjectStore>,
-        connection_pool: ConnectionPool,
-        prover_connection_pool: ConnectionPool,
+        connection_pool: ProverConnectionPool,
```
Comment (Deniallugo) [inline prover/witness_generator/src/basic_circuits.rs:110]: Server?

**Review by Deniallugo — CHANGES_REQUESTED** [review]
It's Titanic job! Well done!
Please review it by yourself one more time a bit more careful.

File: core/lib/db_connection/src/lib.rs [hunk 9]
```diff
@@ -233,35 +293,3 @@ impl ConnectionPool {
...
-            .map(drop)
-            .fetch_optional(storage.conn())
-            .await
-            .unwrap_err();
-        assert_matches!(
-            err,
-            sqlx::Error::Database(db_err) if db_err.message().contains("statement timeout")
-        );
-    }
-}
```
Comment (AnastasiiaVashchuk) [inline core/lib/db_connection/src/lib.rs:267]: moved here - https://github.com/matter-labs/zksync-era/blob/5ab5c9a64eb5212204116ddc87aa132c0b1d7ae4/core/lib/server_dal/src/tests/mod.rs#L307

File: core/lib/server_dal/src/protocol_versions_dal.rs [hunk 10]
```diff
@@ -267,52 +248,4 @@ impl ProtocolVersionsDal<'_, '_> {
...
-                .as_bytes(),
-            vk_commitments.recursion_scheduler_level_vk_hash.as_bytes(),
-        )
-        .fetch_all(self.storage.conn())
-        .await
-        .unwrap()
-        .into_iter()
-        .map(|row| ProtocolVersionId::try_from(row.id as u16).unwrap())
-        .collect()
-    }
```
Comment (AnastasiiaVashchuk) [inline core/lib/server_dal/src/protocol_versions_dal.rs:304]: moved to `prover_dal`

File: core/lib/server_dal/src/protocol_versions_dal.rs [hunk 11]
```diff
@@ -267,52 +248,4 @@ impl ProtocolVersionsDal<'_, '_> {
...
-            "SELECT COUNT(*) as \"count!\" FROM prover_protocol_versions \
-            WHERE id = $1",
-            id as i32
-        )
-        .fetch_one(self.storage.conn())
-        .await
-        .unwrap()
-        .count
-            > 0
-    }
```
Comment (AnastasiiaVashchuk) [inline core/lib/server_dal/src/protocol_versions_dal.rs:317]: moved to `prover_dal`

File: core/lib/zksync_core/src/metadata_calculator/mod.rs [hunk 12]
```diff
@@ -140,15 +141,17 @@ impl MetadataCalculator {
 
     pub async fn run(
         self,
-        pool: ConnectionPool,
-        prover_pool: ConnectionPool,
+        pool: ServerConnectionPool,
+        // short-term change for the prover pool
```
[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply core/lib/zksync_core/src/metadata_calculator/mod.rs:145]: https://linear.app/matterlabs/issue/PLA-677/come-up-with-a-proper-solution-for-running-tree-with-a-prover-pool-in

File: core/lib/server_dal/src/healthcheck.rs [hunk 13]
```diff
@@ -0,0 +1,52 @@
+use serde::Serialize;
```
[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply core/lib/server_dal/src/healthcheck.rs:1]: i thought about it, but not sure how to implement it

**Comment by popzxc** [comment]
Will be tackled as a sequence of smaller PRs

**Pull request closed without merging**